ouroboros-consensus/src/Ouroboros/Consensus/Protocol/HardFork.hs

```diff
@@ -0,0 +1,83 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE TypeFamilies        #-}
+{-# LANGUAGE TypeOperators       #-}
+
+module Ouroboros.Consensus.Protocol.HardFork (
+    CanHardFork (..)
+  , Forked (..)
+  , forked
+  , HardForksTo
+  , NodeConfig (..)
+  ) where
+
+import           Data.Bifunctor (Bifunctor(..))
+
+import           Ouroboros.Network.AnchoredFragment (headHash)
+import           Ouroboros.Network.Block (ChainHash (..))
+
+import           Ouroboros.Consensus.Protocol.Abstract
+import           Ouroboros.Consensus.Util (Empty)
+
+-- | Protocols implementing this class will give up control to a new protocol
+--   when the condition becomes true
+class OuroborosTag p => CanHardFork p where
+  shouldHardFork :: LedgerView p -> Bool
+
+-- | A sum type to represent values before and after a hard fork
+data Forked a b = BeforeFork a | AfterFork b
+  deriving (Eq, Ord, Show)
+
+instance Bifunctor Forked where
+  bimap f g = forked (BeforeFork . f) (AfterFork . g)
+
+forked :: (a -> c) -> (b -> c) -> Forked a b -> c
+forked f g = \case
+  BeforeFork a -> f a
+  AfterFork b -> g b
+
+-- | A protocol that acts as @p1@ until a hard fork switches to @p2@
+data HardForksTo p1 p2
+
+instance
+  (OuroborosTag p1, OuroborosTag p2)
+  => OuroborosTag (p1 `HardForksTo` p2) where
+
+  data NodeConfig (p1 `HardForksTo` p2)
+    = ForkedNodeConfig
+        { nodeConfigBeforeFork :: NodeConfig p1
+        , nodeConfigAfterFork :: NodeConfig p2
+        }
+
+  type NodeState (p1 `HardForksTo` p2) = Forked (NodeState p1) (NodeState p2)
+
+  type ChainState (p1 `HardForksTo` p2) = Forked (ChainState p1) (ChainState p2)
+
+  type IsLeader (p1 `HardForksTo` p2) = Forked (IsLeader p1) (IsLeader p2)
+
+  type LedgerView (p1 `HardForksTo` p2) = Forked (LedgerView p1) (LedgerView p2)
+
+  type ValidationErr (p1 `HardForksTo` p2)
+    = Forked (ValidationErr p1) (ValidationErr p2)
+
+  -- TODO: Figure this one out as we might need some or of constraints
+  type SupportedHeader (p1 `HardForksTo` p2) = Empty
+
+  -- TODO: The intuition here is that we should prefer candidate from before the fork, unless our
+  preferCandidate nodeConfig forkedHeader1 forkedHeader2 = undefined
```

Comment:
Thinking out loud (I haven't tried it out yet, so it may totally not work): can we apply the same pattern as for [`ModChainSel`](https://github.com/input-output-hk/ouroboros-network/blob/fb505ac7f90e2ec2f95d190fd1350e109da3b58f/ouroboros-consensus/src/Ouroboros/Consensus/Protocol/ModChainSel.hs#L27)? Add a method to `CanHardFork`: `beforeOrAfterFork :: HasHeader b => b -> ... -> Forked () ()` (where `...` can be filled in with other arguments that are needed and the two units in `Forked` could be "proofs" of before/after the fork?) which uses the `SlotNo` or `BlockNo` to determine whether the block/header is before or after the fork (maybe `shouldHardFork` can be derived/unified from/with this method). This method can then be used to implement this instance. The advantage is that you only have to implement the `OuroborosTag` instance for `HardForksTo` once.